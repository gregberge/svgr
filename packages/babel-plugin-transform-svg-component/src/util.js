export const getProps = ({ types: t }, opts) => {
  const props = []

  if (opts.ref) {
    props.push(
      t.objectProperty(
        t.identifier('svgRef'),
        t.identifier('svgRef'),
        false,
        true,
      ),
    )
  }

  if (opts.titleProp) {
    props.push(
      t.objectProperty(
        t.identifier('title'),
        t.identifier('title'),
        false,
        true,
      ),
    )

    props.push(
      t.objectProperty(
        t.identifier('titleId'),
        t.identifier('titleId'),
        false,
        true,
      ),
    )
  }

  if (opts.expandProps) {
    props.push(t.restElement(t.identifier('props')))
  }

  if (props.length === 0) {
    return null
  }

  if (props.length === 1 && opts.expandProps) {
    return t.identifier('props')
  }

  return t.objectPattern(props)
}

export const getImport = ({ types: t }, opts) => {
  const importDeclarations = [
    t.importDeclaration(
      [t.importNamespaceSpecifier(t.identifier('React'))],
      t.stringLiteral('react'),
    ),
  ]

  if (opts.native) {
    if (opts.native.expo) {
      importDeclarations.push(t.importDeclaration([], t.stringLiteral('expo')))
    } else {
      importDeclarations.push(
        t.importDeclaration(
          [t.importDefaultSpecifier(t.identifier('Svg'))],
          t.stringLiteral('react-native-svg'),
        ),
      )
    }
  }

  return importDeclarations
}

export const getExport = ({ template }, opts) => {
  let result = ''
  let exportName = opts.state.componentName

  if (opts.memo) {
    const nextExportName = `Memo${exportName}`
    result += `const ${nextExportName} = React.memo(${exportName})\n\n`
    exportName = nextExportName
  }

  if (opts.ref) {
    const nextExportName = `ForwardRef`
    result += `const ${nextExportName} = React.forwardRef((props, ref) => <${exportName} svgRef={ref} {...props} />)\n\n`
    exportName = nextExportName
  }

  if (opts.state.caller && opts.state.caller.previousExport) {
    result += `${opts.state.caller.previousExport}\n`
    result += `export { ${exportName} as ReactComponent }`
    return template.ast(result, {
      plugins: ['jsx'],
    })
  }

  result += `export default ${exportName}`
  return template.ast(result, {
    plugins: ['jsx'],
  })
}
