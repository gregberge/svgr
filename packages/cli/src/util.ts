/* eslint-disable no-console */
import { promises as fs } from 'fs'
import { red } from 'chalk'
import { transform, Config, State } from '@svgr/core'
import svgo from '@svgr/plugin-svgo'
import jsx from '@svgr/plugin-jsx'
import prettier from '@svgr/plugin-prettier'
// @ts-ignore
import camelCase from 'camelcase'
// @ts-ignore
import dashify from 'dashify'

export function transformFilename(
  filename: string,
  filenameCase: string,
): string {
  switch (filenameCase) {
    case 'kebab':
      return dashify(filename.replace(/_/g, '-'), { condense: true })
    case 'camel':
      return camelCase(filename)
    case 'pascal':
      return camelCase(filename, { pascalCase: true })
    default:
      throw new Error(`Unknown --filename-case ${filenameCase}`)
  }
}

export const convert = (
  code: string,
  config: Config,
  state: Partial<State>,
): string => {
  return transform.sync(code, config, {
    ...state,
    caller: {
      name: '@svgr/cli',
      defaultPlugins: [svgo, jsx, prettier],
    },
  })
}

export const convertFile = async (
  filePath: string,
  config: Config = {},
): Promise<string> => {
  const code = await fs.readFile(filePath, 'utf-8')
  return convert(code, config, { filePath })
}

export const exitError = (error: string): never => {
  console.error(red(error))
  process.exit(1)
}

export const politeWrite = (data: string, silent?: boolean): void => {
  if (!silent) {
    process.stdout.write(data)
  }
}

export const formatExportName = (name: string): string => {
  if (/[-]/g.test(name) && /^\d/.test(name)) {
    return `Svg${camelCase(name, { pascalCase: true })}`
  }

  if (/^\d/.test(name)) {
    return `Svg${name}`
  }

  return camelCase(name, { pascalCase: true })
}
