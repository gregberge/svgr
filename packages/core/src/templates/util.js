export const getProps = config => {
  const props = []
  if (config.ref) props.push('svgRef')
  if (config.titleProp) props.push('title')
  if (config.expandProps) props.push('...props')

  if (props.length === 0) return '()'
  if (props.length === 1 && config.expandProps) return 'props'

  return `({ ${props.join(', ')} })`
}
