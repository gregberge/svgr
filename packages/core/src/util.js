import path from 'path'
import { upperFirst, camelCase } from 'lodash'

export const pascalCase = str => upperFirst(camelCase(str))

export function getComponentName(state) {
  if (!state.filePath) return 'SvgComponent'
  const componentName = pascalCase(path.parse(state.filePath).name)
  if (Number.isNaN(parseInt(componentName[0], 10))) return componentName
  return `Svg${componentName}`
}

export function expandState(state) {
  return {
    componentName: state.componentName || getComponentName(state),
    ...state,
  }
}
