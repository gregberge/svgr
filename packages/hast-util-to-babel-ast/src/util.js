/**
 * Determines if the specified string consists entirely of numeric characters.
 *
 * @param {*} [value]
 * @returns {boolean}
 */
export function isNumeric(value) {
  return !Number.isNaN(value - parseFloat(value))
}

/**
 * Convert a hyphenated string to camelCase.
 *
 * @param {string} string
 * @returns {string}
 */
export function hyphenToCamelCase(string) {
  return string.replace(/-(.)/g, (match, chr) => chr.toUpperCase())
}

/**
 * Trim the specified substring off the string. If the string does not end
 * with the specified substring, this is a no-op.
 *
 * @param {string} haystack String to search in
 * @param {string} needle   String to search for
 * @return {string}
 */
export function trimEnd(haystack, needle) {
  return haystack.endsWith(needle)
    ? haystack.slice(0, -needle.length)
    : haystack
}
