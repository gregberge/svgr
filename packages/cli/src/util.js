/* eslint-disable no-console */
import fs from 'fs'
import chalk from 'chalk'
import util from 'util'
import convert, { resolveConfig } from '@svgr/core'

export const readFile = util.promisify(fs.readFile)
export const stat = util.promisify(fs.stat)

export async function convertFile(filePath, { config, ...options } = {}) {
  const code = await readFile(filePath, 'utf-8')
  const rcConfig = await resolveConfig(filePath, config)
  return convert(code, { ...rcConfig, ...options }, { filePath })
}

export function exitError(error) {
  console.error(chalk.red(error))
  process.exit(1)
}
