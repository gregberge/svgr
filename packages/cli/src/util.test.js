import path from 'path'
import { convertFile, transformFilename, CASE, formatExportName } from './util'

const FIXTURES = path.join(__dirname, '../../../__fixtures__')

describe('util', () => {
  describe('#convertFile', () => {
    it('should convert a file', async () => {
      const file = path.join(FIXTURES, 'simple/file.svg')
      const result = await convertFile(file)
      expect(result).toMatchSnapshot()
    })

    it('should support a custom config path', async () => {
      const file = path.join(FIXTURES, 'simple/file.svg')
      const result = await convertFile(file, {
        configFile: '__fixtures__/withSvgrRc/.svgrrc',
      })
      expect(result).toMatchSnapshot()
    })
  })

  describe('#transformFilename', () => {
    it('should transform filename', () => {
      expect(transformFilename('FooBar', CASE.CAMEL)).toBe('fooBar')
      expect(transformFilename('FooBar', CASE.KEBAB)).toBe('foo-bar')
      expect(transformFilename('FooBar', CASE.PASCAL)).toBe('FooBar')

      expect(transformFilename('foo_bar', CASE.CAMEL)).toBe('fooBar')
      expect(transformFilename('foo_bar', CASE.KEBAB)).toBe('foo-bar')
      expect(transformFilename('foo_bar', CASE.PASCAL)).toBe('FooBar')
    })
  })

  describe('#formatExportName', () => {
    it('should ensure a valid export name', () => {
      expect(formatExportName('foo-bar')).toBe('FooBar')
      expect(formatExportName('2foo')).toBe('Svg2foo')
      expect(formatExportName('2foo-bar')).toBe('Svg2FooBar')
    })
  })
})
