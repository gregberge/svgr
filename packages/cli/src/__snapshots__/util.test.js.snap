// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util #convertFile should convert a file 1`] = `
"import * as React from 'react'

function SvgFile(props) {
  return (
    <svg width={48} height={1} {...props}>
      <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
    </svg>
  )
}

export default SvgFile
"
`;

exports[`util #convertFile should support a custom config path 1`] = `
"import * as React from 'react'

function SvgFile(props) {
  return (
    <svg width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 48 1\\" {...props}>
      <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
    </svg>
  )
}

export default SvgFile
"
`;
